crates/ruff/src/commands/graph.rs

```diff
@@ -0,0 +1,159 @@
+use crate::args::{ConfigArguments, GraphArgs};
+use crate::resolve::resolve;
+use crate::{resolve_default_files, ExitStatus};
+use anyhow::Result;
+use log::{debug, warn};
+use ruff_graph::{Direction, ImportMap, ModuleDb, ModuleImports};
+use ruff_linter::warn_user_once;
+use ruff_python_ast::{PySourceType, SourceType};
+use ruff_workspace::resolver::{python_files_in_path, ResolvedFile};
+use rustc_hash::FxHashMap;
+use std::collections::BTreeMap;
+use std::path::Path;
+use std::sync::Arc;
+
+/// Generate an import map.
+pub(crate) fn graph(args: GraphArgs, config_arguments: &ConfigArguments) -> Result<ExitStatus> {
+    // Construct the "default" settings. These are used when no `pyproject.toml`
+    // files are present, or files are injected from outside the hierarchy.
+    let pyproject_config = resolve(config_arguments, None)?;
+
+    // Find all Python files.
+    let files = resolve_default_files(args.files, false);
+    let (paths, resolver) = python_files_in_path(&files, &pyproject_config, config_arguments)?;
+
+    if paths.is_empty() {
+        warn_user_once!("No Python files found under the given path(s)");
+        return Ok(ExitStatus::Success);
+    }
+
+    // Resolve all package roots.
+    let package_roots = resolver
+        .package_roots(
+            &paths
+                .iter()
+                .flatten()
+                .map(ResolvedFile::path)
+                .collect::<Vec<_>>(),
+        )
+        .into_iter()
+        .map(|(path, package)| (path.to_path_buf(), package.map(Path::to_path_buf)))
+        .collect::<FxHashMap<_, _>>();
+
+    // Create a database for each source root.
+    let databases = package_roots
+        .values()
+        .filter_map(|package| package.as_deref())
+        .filter_map(|package| package.parent())
+        .map(Path::to_path_buf)
+        .map(|source_root| Ok((source_root.clone(), ModuleDb::from_src_root(source_root)?)))
+        .collect::<Result<BTreeMap<_, _>>>()?;
```

Comment:
What's the motivation for creating a new db for each source root? 

The caching is local to each database, and creating a new database for each package means that we trade less congestion with reduced cache efficiency. This might be fine if the resolved modules in each package root are distinct but I suspect that the files contain cross-package imports and many stdlib imports that now need to resolved multiple times. 


The recommended way to parallelism in Salsa is to implement a `snapshot` method for your `Database` that creates a shallow clone:

```rust
    #[must_use]
    pub fn snapshot(&self) -> Self {
        Self {
            workspace: self.workspace,
            storage: self.storage.clone(),
            files: self.files.snapshot(),
            system: Arc::clone(&self.system),
        }
    }
```

https://github.com/astral-sh/ruff/blob/ecab04e33851b26cf2e8376be46ca01ce40185e6/crates/red_knot_workspace/src/db.rs#L82-L90


And you can then create a unique database handle for each task:

```rust
        let files = WorkspaceFiles::new(db, self);
        let result = Arc::new(std::sync::Mutex::new(Vec::new()));
        let inner_result = Arc::clone(&result);

        let db = db.snapshot();
        let workspace_span = workspace_span.clone();

        rayon::scope(move |scope| {
            for file in &files {
                let result = inner_result.clone();
                let db = db.snapshot();
                let workspace_span = workspace_span.clone();

                scope.spawn(move |_| {
                    let check_file_span = tracing::debug_span!(parent: &workspace_span, "check_file", file=%file.path(&db));
                    let _entered = check_file_span.entered();

                    let file_diagnostics = check_file(&db, file);
                    result.lock().unwrap().extend(file_diagnostics);
                });
            }
        });

        Arc::into_inner(result).unwrap().into_inner().unwrap()
```

https://github.com/astral-sh/ruff/blob/599103c933e129c6ec9be533c2c563721a58f3bb/crates/red_knot_workspace/src/workspace.rs#L206-L220

Edit: I see below that you already did this. I'm still curious for why we need multiple Dbs. I think what we do here is too granular. At least if I remember how package roots works correctly. I think it doesn't just generate a db for the top-level packages but also for all sub packages. That means that we have two database for `a.b.c`: one for package `a` and another for `b` if `c` is a module and `a` and `a.b` are packages. The result is that we double-resolve modules for `a.b`, depending on whether they come from the `a` or `a.b.` database. 

I also worry that relative imports don't work due to that because I think we might baile out at the `src_root`. 

Can't we generate a single database or a database for each common ancestor of the packages (which ideally is the src directory)?  

If not, can we add some package paths to `SearchPathSettings::extra_paths`?

Reply:
> If not, can we add some package paths to `SearchPathSettings::extra_paths`?

This isn't really "correct" since `extra_paths` take priority over even the stdlib search path in module resolution. But maybe it'll produce the right answer in most cases.

In the real world, if you have multiple independent first-party packages in a workspace (which may or may not depend on each other), you're likely to have them all editably installed into a venv, and red-knot would statically discover all the first-party packages by reading the `pth` files in the `site-packages` directory and add them as search paths that way. But maybe that approach doesn't work here.

Reply:
The source roots here are actually just the root folders. So, like, in Django, it's just a small level of top-level folders:

```python
["/Users/crmarsh/workspace/django", "/Users/crmarsh/workspace/django/tests", "/Users/crmarsh/workspace/django/tests/admin_scripts/custom_templates", "/Users/crmarsh/workspace/django/tests/admin_scripts/custom_templates/project_template", "/Users/crmarsh/workspace/django/tests/i18n/sampleproject", "/Users/crmarsh/workspace/django/tests/migrations/faulty_migrations/namespace", "/Users/crmarsh/workspace/django/tests/migrations/test_fake_initial_case_insensitive"]
```

Reply:
We could use a single database with extra paths -- that's actually the way I did it initially, though Alex said it's not the intended use-case. The downside, I think, is that it would allow those source roots to import files across one another. Maybe that's correct though! I can change it.

Reply:
> The source roots here are actually just the root folders.

If everything's importable from the workspace root, doesn't that imply you can resolve all modules and submodules relative to a single search path, `src_root`, which would point to the workspace root?

Reply:
Alex is obviously the authority when it comes to search paths and I'm probably wrong haha :laughing: 

The intended setup is to use a single database for an application (except VS code where you can have multiple workspaces). That gives the best cache reuse. If that's possible, great. If we aren't hitting any perf issues, that's great too.

Reply:
Looking at the roots for Django that were listed, I think those are all independent roots (as in, they aren't all importable from a single source root), but they can import each other. So I think it would be more efficient to use a single db, by using multiple you are likely repeating work. It's true that the red-knot resolver (because it was built around the more restrictive assumptions of PEP 561 for Python type checkers, rather than a generic "ordered `sys.path`" abstraction) doesn't obviously accommodate this use case, but I actually think extra-paths is a totally fine way to handle it, at least for the Django case.

> This isn't really "correct" since `extra_paths` take priority over even the stdlib search path in module resolution.

I don't understand this comment: all first-party paths (both the main source root and the extra paths) should always take priority over the stdlib in module resolution. According to PEP 561 order, the first two items in the resolution order are extra paths and then the first-party root. So if you have multiple "first party roots", given that they have to go in _some_ resolution order, I don't see how any incorrectness is introduced by putting N-1 of them as extra paths, and whichever one you want last as the main first-party root.

Reply:
> > This isn't really "correct" since `extra_paths` take priority over even the stdlib search path in module resolution.
> 
> I don't understand this comment: all first-party paths (both the main source root and the extra paths) should always take priority over the stdlib in module resolution.

Ugh, you're quite right. Blame it on the holiday haze... yeah, using `extra_paths` should be fine then, in that case. Sorry for causing unnecessary work @charliermarsh :(

Reply:
> In the real world, if you have multiple independent first-party packages in a workspace (which may or may not depend on each other), you're likely to have them all editably installed into a venv, and red-knot would statically discover all the first-party packages by reading the `pth` files in the `site-packages` directory and add them as search paths that way. But maybe that approach doesn't work here.

In many real-world cases (including Django), you'll have little sub-trees of Python code that aren't packaged at all or installed via Python packaging tools, but end up on `sys.path` in some more bespoke way. That's the sort of scenario `extra-paths` is supposed to be for.